sunkit_image/granule.py

```diff
@@ -0,0 +1,298 @@
+                        segmented_image_fixed[mask == 1] = 1.5
+                        bp_count += 1
+    gran_count = len(values) - 1 - bp_count  # Subtract 1 for IG region.
+    return segmented_image_fixed, bp_count, gran_count
+
+
+def kmeans_segment(data):
+    """
+    Uses a k-means clustering algorithm to cluster data, in order to
+    independently verify the scikit-image clustering method.
+
+    Parameters
+    ----------
+    data : `numpy array`
+        Data to be clustered.
+
+    Returns
+    -------
+    labels : `numpy.ndarray`
+        An array of labels with:
+
+        * 0 are granules
+        * 2 are intergranules
+        * 1 are in-between regions
+    """
+    x_size = np.shape(data)[0]
+    y_size = np.shape(data)[1]
+    data_flat = np.reshape(data, (x_size * y_size, 1))
+    labels_flat = KMeans(n_clusters=3, n_init="auto").fit(data_flat).labels_
+    labels = np.reshape(labels_flat, (x_size, y_size))
+    # Make intergranules = 0 and granules = 1.
+    group0_mean = np.mean(data[labels == 0])
+    group1_mean = np.mean(data[labels == 1])
+    group2_mean = np.mean(data[labels == 2])
+    min_index = np.argmin([group0_mean, group1_mean, group2_mean])
+    segmented_map = np.ones(labels.shape)
+    segmented_map[[labels[:, :] == min_index][0]] -= 1
+    return segmented_map
+
+
+def correlation(segment1, segment2):
+    """
+    Compute the correlation of two segmented arrays.
+
+        -1 if the agreement between two arrays is low, 0 otherwise.
+
+        Designed to be used with `segment` and `segment_kmeans` function.
+
+    Parameters
+    ----------
+    segment1 : `numpy.ndarray`
+        Main array to compare the other input array against.
+    segment2 : `numpy.ndarray`
+        Other array, i.e., data segmented using k-means clustering.
+
+    Returns
+    -------
+    [label, confidence] : `list`
+        label : `int`
+            Summarizes the confidence metric:
+            -1: if agreement is low (below 75%)
+            0: otherwise
+        confidence : `float`
+            The numeric confidence metric:
+            Between 0 and 1 where 0 if there is no agreement and 1 if completely agrees.
+    """
+    total_granules = np.count_nonzero(segment1 == 1)
+    total_intergranules = np.count_nonzero(segment1 == 0)
+    if total_granules == 0:
+        raise ValueError("No granules in `segment1`. It is possible the clustering failed.")
+    if total_intergranules == 0:
+        raise ValueError("No intergranules in `segment1`. It is possible the clustering failed.")
+    x_size = np.shape(segment1)[0]
+    y_size = np.shape(segment1)[1]
+    granule_agreement_count = 0
+    intergranule_agreement_count = 0
+    for i in range(x_size):
+        for j in range(y_size):
+            if segment1[i, j] == 1 and segment2[i, j] == 1:
+                granule_agreement_count += 1
+            elif segment1[i, j] == 0 and segment2[i, j] == 0:
+                intergranule_agreement_count += 1

```

Comment:
I have not tested this, but I *think* the following two lines will vectorize this,

```suggestion
    granule_agreement_count = (segment1 == 1 and segment2 == 1).sum()
    intergranule_agreement_count = (segment1 == 0 and segment2 == 0).sum()
```

Please double check that this is indeed the case, but if so, this will be considerably more performant.

Reply:
Hmm.. the only way I was able to make this work was by replacing the `and` with an equality assertion, e.g 
`((segment1 == 1) == (segment2 == 1)).sum()`. Otherwise, Python doesn't like the attempt to compare arrays with `and`. Unfortunately, however, this gives a different result than the original for loops, and I am not quite sure why since I believe it should be doing the same thing.

Reply:
Oops sorry about that.  You're right though `==` will not be the same because that will evaluate to `True` (1) in the case where both conditions are False. I've modified my suggestion above to multiply the two boolean arrays together which (I think!) should correctly return 1 only when both conditions are True since True*False is False and False * False is also False. Thus when you sum, you have a contribution of 1 from only the entries where both contributions are True which is I think equivalent to the for loop.

Reply:
Ah yes, that will do it.

Reply:
There is also the numpy.ma module that provides MaskedArrays that might also be used for this purpose, but I think the simple mask multiplication works fine here.